crates/cheatcodes/src/test/assume.rs

```diff
@@ -20,10 +63,153 @@ impl Cheatcode for assumeCall {
+        let Self { revertData, reverter } = self;
+        assume_no_revert(
+            ccx.state,
+            ccx.ecx.journaled_state.depth(),
+            Some(revertData.to_vec()),
+            false,
+            Some(*reverter),
+        )
+    }
+}
+
+impl Cheatcode for assumeNoPartialRevert_0Call {
+    fn apply_stateful(&self, ccx: &mut CheatsCtxt) -> Result {
+        let Self { revertData } = self;
+        assume_no_revert(
+            ccx.state,
+            ccx.ecx.journaled_state.depth(),
+            Some(revertData.to_vec()),
+            true,
+            None,
+        )
+    }
+}
+
+impl Cheatcode for assumeNoPartialRevert_1Call {
+    fn apply_stateful(&self, ccx: &mut CheatsCtxt) -> Result {
+        let Self { revertData, reverter } = self;
+        assume_no_revert(
+            ccx.state,
+            ccx.ecx.journaled_state.depth(),
+            Some(revertData.to_vec()),
+            true,
+            Some(*reverter),
+        )
+    }
+}
+
+fn assume_no_revert(
+    state: &mut Cheatcodes,
+    depth: u64,
+    reason: Option<Vec<u8>>,
+    partial_match: bool,
+    reverter: Option<Address>,
+) -> Result {
+    ensure!(state.expected_revert.is_none(), ASSUME_EXPECT_REJECT_MAGIC);
+
+    // if assume_no_revert is not set, set it
+    if state.assume_no_revert.is_none() {
+        state.assume_no_revert = Some(AssumeNoRevert { depth, reasons: None, reverted_by: None });
+        // if reason is not none, create a new AssumeNoRevertParams vec
+        if let Some(reason) = reason {
+            state.assume_no_revert.as_mut().unwrap().reasons =
+                Some(vec![AcceptableRevertParameters { reason, partial_match, reverter }]);
+        }
+    } else {
+        // otherwise, ensure that reasons vec is not none and new reason is also not none
+        let valid_assume =
+            state.assume_no_revert.as_ref().unwrap().reasons.is_some() && reason.is_some();
+        ensure!(valid_assume, ASSUME_REJECT_MAGIC);
+        // and append the new reason
+        state
+            .assume_no_revert
+            .as_mut()
+            .unwrap()
+            .reasons
+            .as_mut()
+            .unwrap()
+            .push(AcceptableRevertParameters { reason: reason.unwrap(), partial_match, reverter });
+    }
+
+    Ok(Default::default())
+}
+
+pub(crate) fn handle_assume_no_revert(
+    assume_no_revert: &AssumeNoRevert,
+    status: InstructionResult,
+    retdata: &Bytes,
+    known_contracts: &Option<ContractsByArtifact>,
+    reverter: Option<&Address>,
+) -> Result<()> {
+    // iterate over acceptable reasons and try to match against any, otherwise, return an Error with
+    // the revert data
+    assume_no_revert.reasons.as_ref().map_or_else(
+        || {
+            // todo: fix this hack to get cheatcode name in error message
+            let retdata_str = retdata.to_string();
+            if retdata_str.contains(ASSUME_REJECT_MAGIC) ||
```

Comment:
I'm packing for a trip tomorrow so will take a look tomorrow afternoon, but for now `testMultipleAssumes_ThrowOnGenericNoRevert_fails` fails because `vm.assumeNoRevert` is swallowing the cheatcode revert from a more-specific call to `vm.assumeNoRevert`. `apply_dispatch` seems to modify the revert data (after checking it's a string), so this check is probably failing that comparison.